mitxgraders/helpers/calc.py

```diff
@@ -446,10 +447,18 @@ def parse_algebra(self):
         # Predefine recursive variables.
         expr = Forward()
 
-        # Handle variables passed in. They must start with letters/underscores
+        # Handle variables passed in. They must start with a letter
         # and may contain numbers afterward.
-        inner_varname = Combine(Word(alphas + "_", alphanums + "_") + ZeroOrMore("'"))
-        varname = Group(inner_varname)("variable")
+        inner_varname = Combine(Word(alphas, alphanums + "_") + ZeroOrMore("'"))
+        # Alternative variable name in tensor format
+        # Tensor name must be alphas only, indices may be alphanumeric
+        # e.g., U_{ijk}^{123}
+        indices = Literal("{") + Word(alphanums) + Literal("}")
+        tensor_lower = Combine(Word(alphas) + Literal("_") + indices)
+        tensor_upper = Combine(Word(alphas) + Literal("^") + indices)
+        tensor_mixed = Combine(Word(alphas) + Literal("_") + indices + Literal("^") + indices)
+        # Try for tensor mixed first, then upper, then lower, then generic variable
+        varname = Group(tensor_mixed | tensor_upper | tensor_lower | inner_varname)("variable")
```

Comment:
Do you want to allow primes after a tensor? I think this might be good. Also, if primes are allowed, I think this could be simplified to

```
inner_varname = Combine(Word(alphas, alphanums + "_") + ZeroOrMore("'"))
# Alternative variable name in tensor format
# Tensor name must be alphas only, indices may be alphanumeric
# e.g., U_{ijk}^{123}
upper_indices = Literal("^{") + Word(alphanums) + Literal("}")
lower_indices = Literal("_{") + Word(alphanums) + Literal("}")
tensor = Combine( Word(alphanums) + Optional(lower_indices) + Optional(upper_indices) + ZeroOrMore("'") )
# Try for tensor mixed first, then upper, then lower, then generic variable
varname = Group(tensor | inner_varname)("variable")
```

(In fact, in this case, I believe `tensor` will almost always match. The only case where I believe `inner_varname` would match is variables with multiple  underscores: `a_1_2_3` or `b____4`. I don't see a reason to even allow such variables, except that edX does.)

Reply:
I thought about primed tensors and decided not to, but since you've provided the code... One change I'm making is in the tensor= line, changing Word(alphanums) to Word(alphas, alphanums).